include/contraction/contractionGraph.hpp

```diff
@@ -40,220 +40,304 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+        }
+
+        pgassert(this->is_undirected());
+        BGL_FORALL_OUTEDGES_T(u, e, this->graph, G) {
+            if (this->adjacent(u, e) == v) {
+                contracted_vertices +=
+                    (this->graph[e]).get_contracted_vertices();
+                if ((this->graph[e]).cost < min_cost) {
+                    min_cost = (this->graph[e]).cost;
+                    min_edge = e;
+                    found = true;
+                }
+            }
+        }
+        return std::make_tuple(min_cost, contracted_vertices, found);
+    }
+
+    /*!
+        @brief print the graph with contracted vertices of all vertices and edges
+    */
+    friend
+    std::ostream& operator << (std::ostream &os, const contractionGraph &g) {
+        EO_i out, out_end;
+        for (auto vi = vertices(g.graph).first;
+                vi != vertices(g.graph).second;
+                ++vi) {
+            if ((*vi) >= g.num_vertices()) break;
+            os << g.graph[*vi].id << "(" << (*vi) << ")"
+                << g.graph[*vi].get_contracted_vertices() << std::endl;
+            os << " out_edges_of(" << g.graph[*vi].id << "):";
+            for (boost::tie(out, out_end) = out_edges(*vi, g.graph);
+                    out != out_end; ++out) {
+                os << ' ' << g.graph[*out].id
+                    << "=(" << g.graph[g.source(*out)].id
+                    << ", " << g.graph[g.target(*out)].id << ") = "
+                    <<  g.graph[*out].cost <<"\t";
+            }
+            os << std::endl;
+        }
+        return os;
+    }
+
+    /*!
+        @brief add a shortcut to the graph during contraction
+        @param [in] edge of type *CH_edge* is to be added
+        @param [in] u vertex
+        @param [in] v vertex
+        u -> w -> v
+        u -> v
+        edge (u, v) is a new edge e
+        contracted_vertices = w + contracted vertices
+    */
+    void add_shortcut(const CH_edge &edge, V u, V v) {
+        bool inserted;
+        E e;
+        if (edge.cost < 0) return;
+        boost::tie(e, inserted) = boost::add_edge(u, v, this->graph);
+        this->graph[e]= edge;
+    }
+
+    /*
+     *
+     * u ----e1{v1}----> v ----e2{v2}----> w
+     *
+     * e1: min cost edge from u to v
+     * e2: min cost edge from v to w
+     *
+     * result:
+     * u ---{v+v1+v2}---> w
+     *
+     */
+    void process_shortcut(V u, V v, V w) {
+        auto e1 = get_min_cost_edge(u, v);
+        auto e2 = get_min_cost_edge(v, w);
+
+        if (std::get<2>(e1) && std::get<2>(e2)) {
+            auto contracted_vertices = std::get<1>(e1) + std::get<1>(e2);
+            double cost = std::get<0>(e1) + std::get<0>(e2);
+            contracted_vertices += (this->graph[v]).id;
+            contracted_vertices += (this->graph[v]).get_contracted_vertices();
+
+            // Create shortcut
+            CH_edge shortcut(
+                get_next_id(),
+                (this->graph[u]).id,
```

Comment:
```suggestion
                this->graph[u].id,
```

Parenthesis are useless here